File: 1 - LEETCODE/05 - ALL TOPICS/Problema-I/E0001-Two_Sum/Solucion.js
/*
    ? Problema 0001 - Suma de Dos
    * Autor : Marcos Yana 
*/

// ! Primera Solucion

// inicio leetcode
const twoSum = function (nums, target){
    const numArray = new Map();
    for (let i = 0; i < nums.length; i++) {
        const almacena = nums[i];
        const complemento = target - almacena;
        if(numArray.has(complemento)){
            return [numArray.get(complemento), i];
        }
        numArray.set(almacena, i);
    }
}
//fin leetcode

// ? Verificamos en consola

console.log(twoSum([2, 7, 11, 15], 9)); // [0, 1]
console.log(twoSum([3, 2, 4], 6)); // [0, 1]
console.log(twoSum([3, 3], 6)); // [0, 1]

File: 1 - LEETCODE/05 - ALL TOPICS/Problema-I/E0001-Two_Sum/Solucion.py
"""
    ? Problema 0001 - Suma de Dos
    * Autor : Marcos Yana 
    
"""


# ! Primera Solucion
from typing import List

class Solution:
    def twoSum (self, nums: List[int],target: int) -> List[int]:
        almacena = {}
        for i, x in enumerate(nums):
            complemento = target - x
            if complemento in almacena:
                return [almacena[complemento],i]
            almacena[x] = i
            
# ? Verificamos en consola

solucion = Solution()
print(solucion.twoSum(nums=[2, 7, 11, 15], target=9))
print(solucion.twoSum(nums=[3, 2, 4], target=6))
print(solucion.twoSum(nums=[3, 3], target=6))

